### PR TITLE
GH-38326: [C++][Parquet] check the decompressed page size same as size in page header

### LINKED ISSUE #38326: [C++][Parquet] Compression: check the decompressed page size same as size in page header
### Describe the enhancement requested

Currently, arrow-cpp parquet implemention doesn't check the decompressed page size. This might cause some problems when decompress a weird page.

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

As mentioned in issue, currently we only decompress the page without checking the decompress size. 
This patch add a checkings.

### What changes are included in this PR?

Throw exception when size not matches in `SerializedPageReader::DecompressIfNeeded`

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Non-conforming files may throw an exception while they would silently return invalid results before.

* Closes: #38326

### CODE REVIEW

**Comment by mapleFU** [comment]
cc @wgtmac @pitrou

File: cpp/src/parquet/column_reader.cc [hunk 1]
```diff
@@ -594,10 +594,17 @@ std::shared_ptr<Buffer> SerializedPageReader::DecompressIfNeeded(
...
-      uncompressed_len - levels_byte_len,
-      decompression_buffer_->mutable_data() + levels_byte_len));
+  PARQUET_ASSIGN_OR_THROW(
+      auto decompressed_len,
+      decompressor_->Decompress(compressed_len - levels_byte_len,
+                                page_buffer->data() + levels_byte_len,
+                                uncompressed_len - levels_byte_len,
+                                decompression_buffer_->mutable_data() + levels_byte_len));
+  if (decompressed_len != uncompressed_len - levels_byte_len) {
+    throw ParquetException("Page didn't decompress to expected size, expect: " +
```
Comment (pitrou) [inline cpp/src/parquet/column_reader.cc:604]:
```suggestion
    throw ParquetException("Page didn't decompress to expected size, expected: " +
```

**Comment by pitrou** [comment]
Can you perhaps try reading some third-party Parquet files with this, if you can find some?

**Comment by mapleFU** [comment]
@pitrou When I'm writing test, I found that BROTLI will always return samesize whatever I give to it...

File: cpp/src/parquet/file_deserialize_test.cc [hunk 2]
```diff
@@ -879,6 +879,38 @@ TEST_F(TestPageSerde, DataPageV2CrcCheckNonExistent) {
                          /* write_data_page_v2 */ true);
 }
 
+TEST_F(TestPageSerde, BadCompressedPageSize) {
+  auto codec_types = GetSupportedCodecTypes();
+  const int32_t num_rows = 32;  // dummy value
+  data_page_header_.num_values = num_rows;
+  std::vector<uint8_t> faux_data;
+  test::random_bytes(1024, 0, &faux_data);
+  for (auto codec_type : codec_types) {
```
Comment (mapleFU) [inline cpp/src/parquet/file_deserialize_test.cc:888]: In my local pc, the brotli will "success" decompression.

I've read the code of `BrotliCodec`, but it has a `output_size` and it's used. I didn't dive into it.

Comment (mapleFU) [reply cpp/src/parquet/file_deserialize_test.cc:888]: Ah I've make a mistake and confuse the size.

File: cpp/src/parquet/file_deserialize_test.cc [hunk 3]
```diff
@@ -879,6 +879,45 @@ TEST_F(TestPageSerde, DataPageV2CrcCheckNonExistent) {
...
+    InitSerializedPageReader(num_rows, codec_type);
+
+    EXPECT_THROW_THAT(
+        [&]() { page_reader_->NextPage(); }, ParquetException,
+        ::testing::AnyOf(
+            ::testing::Property(
+                &ParquetException::what,
+                ::testing::HasSubstr("Page didn't decompress to expected size")),
+            // Some decompressor, like zstd, might be able to detect the error
+            // before checking the page size.
```
Comment (mapleFU) [inline cpp/src/parquet/file_deserialize_test.cc:914]: ZSTD would throw an `IOError: Corrupt ...` here.

File: cpp/src/parquet/column_reader.cc [hunk 4]
```diff
@@ -594,10 +594,17 @@ std::shared_ptr<Buffer> SerializedPageReader::DecompressIfNeeded(
...
-      compressed_len - levels_byte_len, page_buffer->data() + levels_byte_len,
-      uncompressed_len - levels_byte_len,
-      decompression_buffer_->mutable_data() + levels_byte_len));
+  PARQUET_ASSIGN_OR_THROW(
+      auto decompressed_len,
+      decompressor_->Decompress(compressed_len - levels_byte_len,
+                                page_buffer->data() + levels_byte_len,
+                                uncompressed_len - levels_byte_len,
+                                decompression_buffer_->mutable_data() + levels_byte_len));
+  if (decompressed_len != uncompressed_len - levels_byte_len) {
```
Comment (wgtmac) [inline cpp/src/parquet/column_reader.cc:603]: The requirement is that uncompressed_len from the page header is something that we can always trust.

Comment (mapleFU) [reply cpp/src/parquet/column_reader.cc:603]: I've seen the arrow-rs implemention, it check the size here. The main problem is that the size is not exact, the decoder will have bad tail buffer, like getting `0` in that multi-gzip file.

File: cpp/src/parquet/file_deserialize_test.cc [hunk 5]
```diff
@@ -879,6 +879,45 @@ TEST_F(TestPageSerde, DataPageV2CrcCheckNonExistent) {
...
 }
 
+TEST_F(TestPageSerde, BadCompressedPageSize) {
+  auto codec_types = GetSupportedCodecTypes();
+  const int data_page_bytes = 8192;
+  const int32_t num_rows = 32;  // dummy value
+  data_page_header_.num_values = num_rows;
+  std::vector<uint8_t> faux_data;
+  // Resize and make all to 1 to make it well-compressed.
+  faux_data.resize(data_page_bytes, 1);
```
Comment (pitrou) [inline cpp/src/parquet/file_deserialize_test.cc:889]: Nit
```suggestion
  // A well-compressible piece of data
  std::vector<uint8_t> faux_data(data_page_bytes, 1);
```

File: cpp/src/parquet/file_deserialize_test.cc [hunk 6]
```diff
@@ -879,6 +879,45 @@ TEST_F(TestPageSerde, DataPageV2CrcCheckNonExistent) {
...
+    const uint8_t* data = faux_data.data();
+    int data_size = static_cast<int>(faux_data.size());
+
+    int64_t max_compressed_size = codec->MaxCompressedLen(data_size, data);
+    buffer.resize(max_compressed_size);
+
+    int64_t actual_size;
+    ASSERT_OK_AND_ASSIGN(
+        actual_size, codec->Compress(data_size, data, max_compressed_size, &buffer[0]));
+    ASSERT_NO_FATAL_FAILURE(WriteDataPageHeader(data_page_bytes, data_size + 1,
```
Comment (pitrou) [inline cpp/src/parquet/file_deserialize_test.cc:902]:
```suggestion
    // Write a data page header declaring a larger decompressed size than actual
    ASSERT_NO_FATAL_FAILURE(WriteDataPageHeader(data_page_bytes, data_size + 1,
```

File: cpp/src/parquet/file_deserialize_test.cc [hunk 7]
```diff
@@ -879,6 +879,45 @@ TEST_F(TestPageSerde, DataPageV2CrcCheckNonExistent) {
                          /* write_data_page_v2 */ true);
 }
 
+TEST_F(TestPageSerde, BadCompressedPageSize) {
+  auto codec_types = GetSupportedCodecTypes();
```
Comment (pitrou) [inline cpp/src/parquet/file_deserialize_test.cc:883]:
```suggestion
  // GH-38326: an exception should be raised if a compressed data page
  // decompresses to a smaller size than declared in the data page header.
  auto codec_types = GetSupportedCodecTypes();
```

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a lot @mapleFU , mostly LGTM.

**Review by pitrou — APPROVED** [review]

**Pull request merged**